pkg/common/metrics/stats_reporter.go

```diff
@@ -0,0 +1,225 @@
+	{regexp.MustCompile(`response-rate`), "Response Rate"},
+	{regexp.MustCompile(`response-size`), "Response Size"},
+	{regexp.MustCompile(`requests-in-flight`), "Requests in Flight"},
+	{regexp.MustCompile(`compression-ratio`), "Compression Ratio"},
+	{regexp.MustCompile(`batch-size`), "Batch Size"},
+	{regexp.MustCompile(`record-send-rate`), "Record Send Rate"},
+	{regexp.MustCompile(`records-per-request`), "Records Per Request"},
+	{regexp.MustCompile(`-for-topic-`), " for Topic "},
+	{regexp.MustCompile(`-for-broker-`), " for Broker "},
+}
+
+// Since regular expressions are somewhat costly and the metrics are repetitive, this cache will hold a simple
+// string-to-string direct replacement
+var replacementCache = map[string]string{}
+
+// Define StatsReporter Structure
+type Reporter struct {
+	views        map[string]*view.View
+	measurements map[string]stats.Measure
+	logger       *zap.Logger
+	tagCtx       context.Context
+}
+
+// StatsReporter Constructor
+func NewStatsReporter(log *zap.Logger) StatsReporter {
+	return &Reporter{
+		views:        make(map[string]*view.View),
+		logger:       log,
+		tagCtx:       context.Background(),
+		measurements: make(map[string]stats.Measure)}
+}
+
+// Our RecordWrapper, which defaults to the knative metrics.Record()
+// This wrapper function facilitates minimally-invasive unit testing of the
+// Report functionality without requiring live servers to be started.
+var RecordWrapper = metrics.Record
+
+// Some type aliases for the otherwise unwieldy metric collection map-of-maps-to-interfaces
+type ReportingItem = map[string]interface{}
+type ReportingList = map[string]ReportingItem
+
+//
+// Report The Sarama Metrics (go-metrics) Via Knative / OpenCensus Metrics
+//
+func (r *Reporter) Report(list ReportingList) {
+
+	// Validate The Metrics
+	if len(list) > 0 {
+
+		// Loop Over The Observed Metrics
+		for metricKey, metricValue := range list {
+
+			// Record Each Individual Metric Item
+			for saramaKey, saramaValue := range metricValue {
+				r.recordMeasurement(metricKey, saramaKey, saramaValue)
+			}
+		}
+	}
+}
+
+// Creates and registers a new view in the OpenCensus context, adding it to the Reporter's known views
+func (r *Reporter) createView(ctx context.Context, measure stats.Measure, name string, description string) context.Context {
+	key, err := tag.NewKey(name)
+	if err != nil {
+		r.logger.Error("Failed To Create New OpenCensus Key For Sarama Metric", zap.String("Key", name), zap.Error(err))
+		return ctx
+	}
+	ctx, err = tag.New(ctx, tag.Insert(key, name))
+	if err != nil {
+		r.logger.Error("Failed To Create New OpenCensus Tag For Sarama Metric", zap.String("Key", name), zap.Error(err))
+		return ctx
+	}
+
+	newView := &view.View{
+		Name:        name,
+		Description: description,
+		Measure:     measure,
+		Aggregation: view.LastValue(),
+	}
+	err = view.Register(newView)
+	if err != nil {
+		r.logger.Error("failed to register opencensus views", zap.Error(err))
```

Comment:
nit: `Failed...` ?

Reply:
https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Reply:
I've seen in this repo that error structs created via mechanisms like errors.New() and fmt.Errorf follow the style guideline of "no initial capital, no trailing punctuation" but not necessarily the direct logger output (fmt.Print, zap.Info, etc).  I don't know if there's a specific style guide for the zap output -- in this particular case the error was copied from eventing code (knative.dev/eventing/pkg/channel/stats_reporter.go) and I didn't think to change it to match the file style.  I would be happy to make things at least consistent (or better, consistently correct) if there is a recommended style.

Reply:
IMO, according to that code conventions page I posted above, your code is right